File: dsa-problems/Arrays/solutions/java/ARR-008-partner-pair-sum-forbidden.java
import java.util.*;

class Solution {
    public boolean hasPairWithForbidden(int[] arr, int target, Set<Integer> forbidden) {
        int left = 0;
        int right = arr.length - 1;

        while (left < right) {
            // Skip forbidden indices
            if (forbidden.contains(left)) {
                left++;
                continue;
            }
            if (forbidden.contains(right)) {
                right--;
                continue;
            }

            // Standard 2-Sum Logic
            long sum = (long)arr[left] + arr[right];

            if (sum == target) {
                return true;
            } else if (sum < target) {
                left++;
            } else {
                right--;
            }
        }

        return false;
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = sc.nextInt();

        int target = sc.nextInt();
        int f = sc.nextInt();
        Set<Integer> forbidden = new HashSet<>();
        for (int i = 0; i < f; i++) forbidden.add(sc.nextInt());

        Solution solution = new Solution();
        boolean result = solution.hasPairWithForbidden(arr, target, forbidden);
        System.out.println(result ? "true" : "false");
        sc.close();
    }
}

File: dsa-problems/Arrays/solutions/ARR-008-partner-pair-sum-forbidden.py
import sys

def has_pair_with_forbidden(arr: list[int], target: int, forbidden: set[int]) -> bool:
    left = 0
    right = len(arr) - 1

    while left < right:
        # Skip forbidden indices on the left
        while left < right and left in forbidden:
            left += 1
        # Skip forbidden indices on the right
        while left < right and right in forbidden:
            right -= 1

        # Check if we still have valid pointers
        if left >= right:
            break

        current_sum = arr[left] + arr[right]

        if current_sum == target:
            return True
        elif current_sum < target:
            left += 1
        else:
            right -= 1

    return False

def main():
    n = int(input())
    arr = list(map(int, input().split()))
    target = int(input())
    f = int(input())
    forbidden = set(map(int, input().split())) if f > 0 else set()

    result = has_pair_with_forbidden(arr, target, forbidden)
    print("true" if result else "false")

if __name__ == "__main__":
    main()

